src/main/java/org/folio/circulation/services/FeeFineService.java

```diff
@@ -1,58 +1,164 @@
+        .withAmount(createdAccount.getAmount())
+        .withUserId(createdAccount.getUserId())
+        .withAction(createdAccount.getFeeFineType())
+        .withAccountId(createdAccount.getId())
+        .withCreatedAt(pair.getRight().getName())
+        .withCreatedBy(pair.getLeft())
+        .build()))
+      .thenCompose(r -> r.after(feeFineActionRepository::create))
+      .thenApply(r -> r.map(notUsed -> null));
+  }
+
+  public CompletableFuture<Result<Void>> refundAndCloseAccounts(List<RefundAccountCommand> accounts) {
+    return allOf(accounts.stream()
+      .map(this::refundAndCloseAccount)
+      .toArray(CompletableFuture[]::new))
+      .thenApply(Result::succeeded)
+      .exceptionally(CommonFailures::failedDueToServerError);
+  }
+
+  private CompletableFuture<Result<Void>> refundAndCloseAccount(RefundAccountCommand refund) {
+    return fetchServicePointAndUser(refund.getStaffUserId(), refund.getServicePointId())
+      .thenApply(r -> r.map(pair -> new AccountRefundContext(refund.getAccountToRefund())
+        .withServicePoint(pair.getRight())
+        .withUser(pair.getLeft())))
+      .thenCompose(r -> r.after(this::processRefundAndClose));
+  }
+
+  private CompletableFuture<Result<Void>> processRefundAndClose(AccountRefundContext context) {
+    final Result<AccountRefundProcessor> processorResult = getProcessor(context.getAccount());
+    if (processorResult.failed()) {
+      return completedFuture(failed(processorResult.cause()));
+    }
+
+    final AccountRefundProcessor processor = processorResult.value();
+
+    return createRefundAndCloseActions(processor, context)
+      .thenCompose(r -> r.after(notUsed -> updateAccount(processor, context)));
+  }
+
+  private CompletableFuture<Result<Void>> updateAccount(
+    AccountRefundProcessor processor, AccountRefundContext context) {
+
+    final Account account = context.getAccount();
+    if (account.hasTransferredAmount()) {
+      processor.onTransferAmountRefundActionSaved(context);
+    }
+
+    if (account.hasPaidAmount()) {
+      processor.onPaidAmountRefundActionSaved(context);
+    }
+
+    if (account.hasRemainingAmount()) {
+      processor.onRemainingAmountActionSaved(context);
+    }
+
+    return accountRepository.update(StoredAccount.fromAccount(context.getAccount()));
+  }
+
+  private CompletableFuture<Result<Void>> createRefundAndCloseActions(
+    AccountRefundProcessor processor, AccountRefundContext context) {
+
+    final Account account = context.getAccount();
+    if (account.hasTransferredAmount()) {
+      processor.onHasTransferAmount(context);
+    }
+
+    if (account.hasPaidAmount()) {
+      processor.onHasPaidAmount(context);
+    }
+
+    if (account.hasRemainingAmount()) {
+      processor.onHasRemainingAmount(context);
+    }
+
+    return feeFineActionRepository.createAll(context.getActions());
+  }
+
+  private CompletableFuture<Result<Pair<User, ServicePoint>>> fetchServicePointAndUser(
```

Comment:
Why are these being fetched together? Does the user relate to the service point in some way?

Reply:
No, they don't. It is some reference data needed to build an account action record.

Reply:
Why can they not be fetched separately, is it to make it convenient to retain them whilst moving on to the next step?

Reply:
Extracted to be fetched separately.